storagestate/storagestate.go

```diff
@@ -0,0 +1,783 @@
+			return "", 0, nil, aoserrors.Wrap(err)
+		}
+
+		return stateInstanceFolder, stateQuota, checksum, nil
+	}
+
+	if stateInstanceFolder == "" {
+		stateInstanceFolder = filepath.Join(storageState.stateDir, instanceID)
+		if err = createFolder(stateInstanceFolder, params.UID, params.GID); err != nil {
+			return "", 0, nil, aoserrors.Wrap(err)
+		}
+	}
+
+	setupQuota := storageState.prepareQuota(params.StateQuota, params.StorageQuota)
+
+	if err = SetUserFSQuota(
+		storageState.stateDir, setupQuota, uint32(params.UID), uint32(params.GID)); err != nil {
+		return "", 0, nil, aoserrors.Wrap(err)
+	}
+
+	if checksum, err = storageState.startStateWatching(
+		instanceID, params, stateInstanceFolder, stateChecksum); err != nil {
+		return "", 0, nil, aoserrors.Wrap(err)
+	}
+
+	return stateInstanceFolder, params.StateQuota, checksum, nil
+}
+
+func (storageState *StorageState) prepareQuota(current, add uint64) uint64 {
+	quota := current
+
+	if storageState.isSamePartition {
+		quota += add
+	}
+
+	return quota
+}
+
+func checkInstanceDirExists(instanceDir string) (dir string, err error) {
+	if instanceDir != "" {
+		if _, err = os.Stat(instanceDir); err != nil {
+			if !os.IsNotExist(err) {
+				return "", aoserrors.Wrap(err)
+			}
+
+			log.Warnf("Folder %s doesn't exist", instanceDir)
+
+			return "", nil
+		}
+	}
+
+	return instanceDir, nil
+}
+
+func createStateFileIfNotExist(path string, uid, gid int) error {
+	if _, err := os.Stat(path); err != nil {
+		if !os.IsNotExist(err) {
+			return aoserrors.Wrap(err)
+		}
+
+		if err := createStateFile(path, uid, gid); err != nil {
+			return aoserrors.Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+func createFolder(path string, uid, gid int) error {
+	if err := os.MkdirAll(path, 0o755); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if err := os.Chown(path, uid, gid); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	upperDir := filepath.Join(path, upperDirName)
+
+	if err := os.MkdirAll(upperDir, 0o755); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if err := os.Chown(upperDir, uid, gid); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	workDir := filepath.Join(path, workDirName)
+
+	if err := os.MkdirAll(workDir, 0o755); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if err := os.Chown(workDir, uid, gid); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
+
+func checkAndCreateDir(dir string) error {
+	if _, err := os.Stat(dir); err != nil {
+		if !os.IsNotExist(err) {
+			return aoserrors.Wrap(err)
+		}
+
+		if err = os.MkdirAll(dir, 0o755); err != nil {
+			return aoserrors.Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+func createStateFile(path string, uid, gid int) (err error) {
+	file, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0o600)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+	defer file.Close()
+
+	if err = os.Chown(path, uid, gid); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
+
+func removeDir(dir string) {
```

Comment:
use os.RemoveAll in place, also check and display error.

Reply:
Done